### PR TITLE
#1351 - Support for user-defined kernels that operate on DoFs

### LINKED ISSUE #1351: [LFRic] Support for user-defined kernels that operate on DoFs
Originally we had preliminary support for user-defined kernels that operate on DoFs. However, it produced incorrect code and we removed it until the need arises.

There are several LFRic PSyKAl-lite routines that could be removed with this support, hence this issue. Furthermore, it seems that we need to differentiate between `operates_on = DOF` and `operates_on = DOF_COLUMN`.

### PR DESCRIPTION
**Pull request by oakleybrunt**
Closes #1351

### CODE REVIEW

File: doc/user_guide/dynamo0p3.rst [hunk 1]
```diff
@@ -973,6 +973,59 @@ on a ``CELL_COLUMN`` without CMA Operators. Specifically:
...
+The list of rules for DoF Kernels is as follows:
+
+1) A DoF Kernel must have at least one argument that is a field.
+   This rule reflects that a Kernel operates on some subset of the
+   whole domain and is therefore designed to be called from within
+   a loop that iterates over those subsets of the domain. Only fields
+   are accepted for a Kernel operating on DoFs because fields are the
+   only argument type that contain DoFs (i.e. field vectors are not allowed).
+
+2) All Kernel arguments must be either fields or scalars (`real-`` and/or
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:997]: One too many back ticks here.

File: doc/user_guide/dynamo0p3.rst [hunk 2]
```diff
@@ -973,6 +973,59 @@ on a ``CELL_COLUMN`` without CMA Operators. Specifically:
...
+   `integer`-valued). DoF Kernels cannot accept operators.
+
+3) All field arguments to a given DoF Kernel must be on the same function
+   space. This is because all fields should have the same number.
+   It also means that PSyclone can determine the number of DoFs uniquely when
+   writing to a scalar.
+
+4) They must have at least one modified (i.e. writen to) argument. Unlike
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:1005]: "...to) field argument. Unlike..."

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for this Oakley. I think you've mostly got the top-level rules sorted but you're missing a section on the rules for the kernel subroutine itself - see inline. Note I've brought the branch up to master so you'll need to `pull` before you make changes.

File: doc/user_guide/dynamo0p3.rst [hunk 3]
```diff
@@ -991,40 +991,26 @@ The list of rules for DoF Kernels is as follows:
...
+   space so they have the same number of DoFs.
 
-4) They must have at least one modified (i.e. writen to) argument. Unlike
+4) They must have at least one modified (i.e. writen to) field argument. Unlike
    built-ins, this is not limited and more than one modified argument is
    allowed.
 
 5) A Kernel may not write to a scalar argument. (Only built-ins are permitted
    to do this.) Any scalar arguments must therefore be declared in the metadata
    as `GH_READ` - see :ref:`below<dynamo0-3-kernel-valid-access>`
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:1009]: This link target isn't right. Probably should be "dynamo0.3-kernel-valid-access"? You can check this by doing:

    cd PSyclone
    pip install -e .[doc]
    cd doc/user_guide
    make html

and then check for any errors or warnings.

File: doc/user_guide/dynamo0p3.rst (hunk 2 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: "written" - sorry missed that first time round.

File: doc/user_guide/dynamo0p3.rst [hunk 4]
```diff
@@ -848,7 +848,7 @@ types.
    an ``integer``-valued field as an argument.
 
 .. _lfric-no-cma-mdata-rules:
-   
+
 Rules specific to General-Purpose Kernels without CMA Operators
 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:854]: I've realised that the introductory paragraph to the `Kernel` section:
image
needs updating as we are going to add support for a new kernel type here. When you do that, please could you also add a note that support for this kernel type is to be implemented (and reference the issue). Please also do that in the two new sections you've added about the rules for this kernel type.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks Oakley. Looking pretty good now. I've realised we're missing the rather crucial `OPERATES_ON` for this kernel type and I found another bit of the docs that need updating.
We also need to be clear that support for this new kernel type is not yet implemented (a slight disadvantage of doing things this way but the overhead shouldn't be too bad).

When this is "ready for review" again, please could you add the appropriate label to the PR - that makes it easy for me to see in the list of PRs.

File: doc/user_guide/dynamo0p3.rst (hunk 3 above)
Comment (oakleybrunt) [reply doc/user_guide/dynamo0p3.rst:1009]: thank you :)

File: doc/user_guide/dynamo0p3.rst [hunk 5]
```diff
@@ -772,15 +772,15 @@ Kernel
 
 The general requirements for the structure of a Kernel are explained
 in the :ref:`kernel-layer` section. In the LFRic API there are five
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst]: s/five/six/

File: doc/user_guide/dynamo0p3.rst [hunk 6]
```diff
@@ -772,15 +772,15 @@ Kernel
...
+other five user-supplied kernel types and then goes on to describe
 their metadata and subroutine arguments.
 
-Domain kernels are distinct from the other three, user-supplied kernel
-types in that they must be passed data for the whole domain rather
-than a single cell-column. This permits the use of kernels that have
+Domain kernels are distinct from the other four user-supplied kernel
+types because they must be passed data for the whole domain rather
+than a single cell-column or field. This permits the use of kernels that have
 not been written to conform to the single-column approach which
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst]: "single-column/dof approach..."

File: doc/user_guide/dynamo0p3.rst (hunk 5 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: As there are just a couple of typos, I'll take the liberty of fixing them.

File: doc/user_guide/dynamo0p3.rst [hunk 7]
```diff
@@ -772,15 +772,15 @@ Kernel
...
-other four, user-supplied kernel types and then goes on to describe
+other five user-supplied kernel types and then goes on to describe
 their metadata and subroutine arguments.
 
-Domain kernels are distinct from the other three, user-supplied kernel
-types in that they must be passed data for the whole domain rather
-than a single cell-column. This permits the use of kernels that have
+Domain kernels are distinct from the other four user-supplied kernel
+types because they must be passed data for the whole domain rather
+than a single cell-column or field. This permits the use of kernels that have
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: Fixed.

File: doc/user_guide/dynamo0p3.rst (hunk 6 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: Fixed.

**Review by arporter — APPROVED** [review]
Thanks for making those changes. I've taken the liberty of fixing a couple of typos and this is now ready to go once the CI has passed.

**Pull request merged**